Support speech recognition with whisper models and seq2seq. (#704)
huggingface/optimum-habana: examples/speech-recognition/README.md

# Automatic Speech Recognition Examples

## Table of Contents

- [Automatic Speech Recognition with CTC](#connectionist-temporal-classification)

> 	- [Single GPU example](#single-hpu-ctc)

Reviewer: ```suggestion
	- [Single HPU example](#single-hpu-ctc)
```

> 	- [Multi GPU example](#multi-hpu-ctc)

Reviewer: ```suggestion
	- [Multi HPU example](#multi-hpu-ctc)
```

# Automatic Speech Recognition Examples

## Table of Contents

- [Automatic Speech Recognition with CTC](#connectionist-temporal-classification)
	- [Single GPU example](#single-gpu-ctc)
	- [Multi GPU example](#multi-gpu-ctc)
	- [Examples](#examples-ctc)
		- [TIMIT](#timit-ctc)
		- [Librispeech](#librispeech-ctc)
		- [Common Voice](#common-voice-ctc)
		- [Multilingual Librispeech](#multilingual-librispeech-ctc)
- [Automatic Speech Recognition with Sequence-to-Sequence](#sequence-to-sequence)
	- [Whisper Model](#whisper-model)
	- [Fine tuning](#single-hpu-whisper-training-with-seq2seq)

> 	- [Inference](#single-gpu-seq2seq-inference)

Author: When you say the ones of Transformers example, which one specifically do you mean?
I tried to form a Table of contents, because the transformers repo had one, but may not match what I put here.

> 	- [Inference](#single-gpu-seq2seq-inference)

Reviewer: I mean that the table of content contains several occurrences of "GPU" while it should probably be HPU

> 	- [Inference](#single-gpu-seq2seq-inference)

Author: I will fix this and run the "make style" that you requested.

> 	- [Inference](#single-gpu-seq2seq-inference)

Author: This is done.

    --gaudi_config_name="Habana/wav2vec2" \
    --bf16
```

> ## Sequence to Sequence

Author: I think the above sections are for CTC including the Deepspeed and Inference sections above. The sections below here are for Seq2Seq.

> ## Sequence to Sequence

Reviewer: You're right, but my point is rather that running the seq2seq example with DeepSpeed or for doing only inference can be done the exact same way as the CTC example. But that's fine, this is a nit!

The script [`run_speech_recognition_seq2seq.py`](https://github.com/huggingface/optimum-habana/examples/speech-recognition/run_speech_recognition_seq2seq.py) can be used to fine-tune any [Whisper Sequence-to-Sequence Model](https://huggingface.co/docs/transformers/main/en/model_doc/whisper#whisper) for automatic speech 
recognition on one of the well known speech recognition datasets similar to shown below or a custom dataset. Examples of two datasets using the Whisper model from OpenAI are included below.

### Whisper Model
We can load all components of the Whisper model directly from the pretrained checkpoint, including the pretrained model weights, feature extractor and tokenizer. We simply have to specify our fine-tuning dataset and training hyperparameters.
```bash
# Set model dir and datasets dir as appropriate
export MODEL_DIR=$HOME/huggingface/hub
export DATASETS_DIR=$HOME/huggingface/datasets

# Run this from the optimum-habana repository
cd examples/speech-recognition
# install optimum-habana 
pip install -r requirements.txt

> ```

Reviewer: I don't think we need this

Author: Yes this can be removed.

#### Single HPU Whisper Fine tuning with Seq2Seq

> The following example shows how to fine-tune the [Whisper small](https://huggingface.co/openai/whisper-small) checkpoint on the Hindi subset of [Common Voice 11](https://huggingface.co/datasets/mozilla-foundation/common_voice_11_0) using a single GPU device in half-precision:

Reviewer: ```suggestion
The following example shows how to fine-tune the [Whisper small](https://huggingface.co/openai/whisper-small) checkpoint on the Hindi subset of [Common Voice 11](https://huggingface.co/datasets/mozilla-foundation/common_voice_11_0) using a single HPU device in bf16 precision:
```

```bash
python ../gaudi_spawn.py \
 	--world_size 8 --use_mpi run_speech_recognition_seq2seq.py \
	--model_name_or_path="openai/whisper-large" \
	--dataset_name="mozilla-foundation/common_voice_11_0" \
	--dataset_config_name="hi" \
	--language="hindi" \
	--train_split_name="train+validation" \
	--eval_split_name="test" \
	--gaudi_config_name="gaudi_config.json" \
	--max_steps="5000" \
	--output_dir="./results/whisper-large-hi" \
	--per_device_train_batch_size="16" \
	--gradient_accumulation_steps="2" \
	--per_device_eval_batch_size="16" \
	--logging_steps="25" \
	--learning_rate="1e-5" \
	--warmup_steps="500" \
	--evaluation_strategy="steps" \
	--eval_steps="1000" \
	--save_strategy="steps" \
	--save_steps="1000" \
	--generation_max_length="225" \
	--preprocessing_num_workers="16" \
	--length_column_name="input_length" \
	--max_duration_in_seconds="30" \
	--text_column_name="sentence" \
	--freeze_feature_encoder="False" \
	--gradient_checkpointing \
	--group_by_length \
	--bf16 \
	--overwrite_output_dir \
	--do_train \
	--do_eval \
	--predict_with_generate \
	--use_habana \
	--use_hpu_graphs_for_inference
```

> #### Single GPU Seq2Seq Inference

Reviewer: ```suggestion
#### Single HPU Seq2Seq Inference
```

> The following example shows how to do inference with the [Whisper small](https://huggingface.co/openai/whisper-small) checkpoint on the Hindi subset of [librispeech_asr_demo](https://huggingface.co/datasets/librispeech_asr_demo) using 1 HPU devices in half-precision:

Reviewer: ```suggestion
The following example shows how to do inference with the [Whisper small](https://huggingface.co/openai/whisper-small) checkpoint on the Hindi subset of [librispeech_asr_demo](https://huggingface.co/datasets/librispeech_asr_demo) using 1 HPU device in half-precision:
```

```bash
python run_speech_recognition_seq2seq.py \
    --model_name_or_path="openai/whisper-small" \
	--dataset_name="mozilla-foundation/common_voice_11_0" \
	--dataset_config_name="hi" \
	--language="hindi" \
	--eval_split_name="test" \
 	--gaudi_config_name="gaudi_config.json" \
	--max_steps="5000" \
	--output_dir="./results/whisper-small-clean" \
	--gradient_accumulation_steps="2" \
	--per_device_eval_batch_size="16" \
	--logging_steps="25" \
	--learning_rate="1e-5" \
	--warmup_steps="500" \
	--evaluation_strategy="steps" \
	--eval_steps="1000" \
	--save_strategy="steps" \
	--save_steps="1000" \
	--generation_max_length="225" \
	--preprocessing_num_workers="16" \
	--length_column_name="input_length" \
	--max_duration_in_seconds="30" \
	--text_column_name="sentence" \
	--freeze_feature_encoder="False" \
	--gradient_checkpointing \
	--group_by_length \
	--bf16 \
	--overwrite_output_dir \
	--do_eval \
	--predict_with_generate \
	--use_habana \
	--use_hpu_graphs_for_inference

> ```

Reviewer: Let's keep only one example for inference, whichever you prefer

Author: Ok. Let's keep the 2nd one which uses the common_voice_11_0 dataset.
